### PR TITLE
Clarify `get_pow_block` block-not-found case

### PR DESCRIPTION
**Pull request by hwwhww**
### Issue
Discussed in #2636 
It is possible that the requested block hash is not found via `get_pow_block`.

### How did I fix it

- Make `get_pow_block` returns `Optional[PowBlock]`
- If it returns `None`, the operation is invalid.
- Add some notes about the edge case.

### CODE REVIEW

File: specs/merge/fork-choice.md [hunk 1]
```diff
@@ -64,7 +64,8 @@ class PowBlock(Container):
 
 ### `get_pow_block`
 
-Let `get_pow_block(block_hash: Hash32) -> PowBlock` be the function that given the hash of the PoW block returns its data.
+Let `get_pow_block(block_hash: Hash32) -> Optional[PowBlock]` be the function that given the hash of the PoW block returns its data.
+It may result in `None` if the requested block is not found if execution engine is still syncing.
```
Comment (djrtwo) [inline specs/merge/fork-choice.md:68]: The consensus spec assumes a 1:1 relationship between CL and EL sync. I'm not sure we should put notes about "syncing" in here. From a consensus spec standpoint, it is unknown if a block is a valid transition block if `get_pow_block` cannot be resolved.

*In practice*, CL will process past the transition point even if EL is syncing, and only later know if such a chain was valid or invalid once EL is done syncing. I'm worried the `assert`s in `on_block` imply too heavily a failure

Comment (djrtwo) [reply specs/merge/fork-choice.md:68]: I might be wavering on my resolve here. I don't think we should mention "execution engine" but you could definitely get a block with a terminal block reference that you can't find. Thus there should be an assertion here

Comment (hwwhww) [reply specs/merge/fork-choice.md:68]: @djrtwo

is it okay to shorten to:
```suggestion
It may result in `None` if the requested block is not found.
```

Comment (djrtwo) [reply specs/merge/fork-choice.md:68]: I edited the above. What do you think?

File: specs/merge/fork-choice.md [hunk 2]
```diff
@@ -110,8 +117,12 @@ def on_block(store: Store, signed_block: SignedBeaconBlock) -> None:
 
     # [New in Merge]
     if is_merge_block(pre_state, block.body):
+        # Note: unavailable PoW block(s) may later become available.
+        # Nodes should queue such beacon blocks for later processing.
```
Comment (mkalinin) [inline specs/merge/fork-choice.md:121]: This case is a bit more tricky and we should think what would be a better way of specifying it. For the case when CL and EL are in sync with the network, current specification works. But for the case when they are syncing we might want to add more clarity to the spec.

Queueing these blocks could raise a deadlock in an optimistic sync process. Suppose, the optimistic sync is happening and EL uses state sync to catch up with the head and the network passed the transition point a while ago, so the state for transition block is no longer available due to state trie pruning. By queueing beacon blocks at this point, CL will stuck and won't inform EL about the most recent head that is available in the network.

The right behaviour during the sync stage would be as follows:
- process transition block optimistically without terminal block verification
- once fully synced, get back to the transition block and check if it's been finalized or not
- verify terminal block if the transition block hasn't been finalized
- do re-org if the terminal block is invalid 

We also may add this clarity in a separate PR.

**Review by mkalinin — COMMENTED** [review]
We might want to add more clarity around transition block in the context of optimistic sync. But this could be done in a separate PR

File: specs/merge/fork-choice.md (hunk 2 above)
Comment (djrtwo) [reply specs/merge/fork-choice.md:121]: I consider the optimistic sync process as independent from the consensus specs.

"queuing blocks for later processing" in optimistic beacon chain sync means that you will eventually go back to fully validate them later. The architecture of CL/EL split client is independent of core consensus specs and I want to find the right balance here to not taint the specs with implementation details.

That said, I could see how this is confusing. But I do think that more notes about optimistic processing can/should be included in the Engine API as that is where this implemntation specific architecture design lives

**Review by djrtwo — APPROVED** [review]
lgtm

**Pull request merged**